New extensions: Objectify and Remote Logging

freddyboucher opened the pull request:
This pull request introduces 2 new extensions:
- Objectify
- Remote Logging

## Objectify
This extension contains serializer and deserializer for [Objectify](https://github.com/objectify/objectify) types like **Ref** or **Key**.

To use it add the library to your classpath (for maven, the artifactId is `gwt-jackson-objectify`):
1. Client-side: add `<inherits name="com.github.nmorel.gwtjackson.objectify.GwtJacksonObjectify" />` to your module descriptor XML file.
2. Server-side: register the Jackson module `ObjectifyJacksonModule`

Important note: **Ref\<T\>** is serialized as **DeadRef\<T\>** and the T value is serialized if it is loaded into the session.

## Remote Logging
This extension contains serializer for Throwable.

It allows you to support the equivalent of GWT RPC Remote Logging http://www.gwtproject.org/doc/latest/DevGuideLogging.html#Remote_Logging

To use it add the library to your classpath (for maven, the artifactId is `gwt-jackson-remotelogging`):
1. Client-side: add `<inherits name="com.github.nmorel.gwtjackson.remotelogging.GwtJacksonRemoteLogging" />` to your module descriptor XML file.
2. Server-side: register the Jackson module `RemoteLoggingJacksonModule`


Both extensions are used in the following demo project: [gwt-storage-objectify](https://github.com/freddyboucher/gwt-storage-objectify)

freddyboucher commented:
@nmorel 
After months of development, here it is, extensions are ready.
Could you by any change review it?
Thanks.

nmorel commented:
Hi :)
Thanks for the PR!
That's a lot of code to review, I'll try to review it by the end of the week.

The biggest issue will be the publication. I no longer have an environment to deploy it.

freddyboucher commented:
@nmorel 
Don't worry, there is no rush.
Thanks

nmorel commented on extensions/objectify/pom.xml:
```diff
@@ -0,0 +1,151 @@
+          </execution>
+        </executions>
+      </plugin>
+      <plugin>
```
Do you need 1.7 ? 
Can you try by just setting the following properties on top the files (like in pom examples) : 
```xml
    <maven.compiler.source>1.7</maven.compiler.source>
    <maven.compiler.target>1.7</maven.compiler.target>
```

freddyboucher replied:
Sure!
Coming in few minutes

freddyboucher replied:
And no, it doesn't require 1.7, so I will just remote them

freddyboucher replied:
@nmorel 
I want to keep the PR clean, there are right now 3 commits in this PR:
- make generic key works
- Create the Objectify extension
- Create the Remote Logging extension

Is it OK for you if I amend commits to keep only these 3 commits?

freddyboucher replied:
Fixed!

nmorel replied:
Yes sure. You can rebase or squash :)

nmorel commented on extensions/objectify/src/main/java/com/github/nmorel/gwtjackson/objectify/client/KeyJsonDeserializer.java:
```diff
@@ -0,0 +1,35 @@
+        return INSTANCE;
+    }
+
+    public static <T> KeyJsonDeserializer<T> getInstance() {
```
this one is used ?

freddyboucher replied:
Yes it's used in RefJsonDeserializer#L36

freddyboucher commented:
@nmorel 
Done, I (forced) pushed my changes. It's now Java 1.6 compliant

freddyboucher commented:
@nmorel 
I can't believe a didn't see the failing tests. Here the fix: https://github.com/nmorel/gwt-jackson/pull/129
Should be great to setup Travis to build and run tests on each PR.

freddyboucher commented on extensions/objectify/src/main/java/com/github/nmorel/gwtjackson/objectify/server/ObjectifyJacksonModule.java:
```diff
@@ -0,0 +1,33 @@
+import com.googlecode.objectify.Key;
+import com.googlecode.objectify.Ref;
+
+public class ObjectifyJacksonModule extends SimpleModule {
```
It would mean 2 modules (one client, one server) for each extension?
Would you prefer? up to you...

nmorel replied:
No it's ok. Reading the rest of the code, it would be too long to separate for little gain.

nmorel commented:
Yea I had a Jenkins running tests but my provider stopped OSS support.

freddyboucher commented:
@nmorel 
Thanks for the review

Fixed!
- added Copyright
- use shared **Tester to ensure Client and Server serialize/deserialize same values (I like this approach)

Ready for the second round.

freddyboucher commented:
@nmorel 
Any new comments

freddyboucher commented:
@nmorel 
Oops I close it by mistake

freddyboucher commented:
@nmorel 
Any other comments for this PR?
Thanks

freddyboucher commented:
@nmorel 
As you see (build failed), it's not ready to review it yet.

nmorel commented:
Yes, just saw that. Tell me when it's ok :)

freddyboucher commented:
@nmorel 
I fixed the build, it's now ready to be review again.
Thanks

freddyboucher commented:
@nmorel 
Any chance you could review it this week? (And push it to Maven? :) )
Thanks

nmorel commented:
I merged it. I'll see to deploy a snapshot version.  
It'll take longer for a release as I no longer have an environment to sign the jars.

freddyboucher commented:
@nmorel 
Thanks a lot.

nmorel commented:
@freddyboucher I released a 0.15.0 version. Can you test it please ?
And thanks for your work :)

freddyboucher commented:
@nmorel 
Works like a charm.
Thanks a lot.
